src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+
+        /// Invoked when the table index block and all content blocks have been written.
+        callback: *const fn (*RepairTable, RepairTableResult) void,
+        /// "next" belongs to the `faulty_tables` FIFO.
+        next: ?*RepairTable = null,
+    };
+
+    pub const Options = struct {
+        /// Lower-bound for the limit of concurrent enqueue_block()'s available.
+        blocks_max: usize,
+        /// Maximum number of concurrent enqueue_table()'s.
+        tables_max: usize,
+    };
+
+    options: Options,
+
+    /// Invariants:
+    /// - For every block address in faulty_blocks, ¬free_set.is_free(address).
+    faulty_blocks: FaultyBlocks,
+    /// Index within `faulty_blocks`, used to cycle through block-repair requests.
+    ///
+    /// Invariants:
+    /// - faulty_blocks.count() > 0 implies faulty_blocks_repair_index < faulty_blocks.count()
+    /// - faulty_blocks.count() = 0 implies faulty_blocks_repair_index = faulty_blocks.count()
+    faulty_blocks_repair_index: usize = 0,
+
+    /// Invarants:
+    /// - enqueued_table_blocks + enqueued_single_blocks = faulty_blocks.count()
+    /// - enqueued_table_blocks ≤ options.tables_max * lsm_table_content_blocks_max
+    enqueued_single_blocks: usize = 0,
+    enqueued_table_blocks: usize = 0,
+
+    /// Invariants:
+    /// - For every index address in faulty_tables: ¬free_set.is_free(address).
+    faulty_tables: FIFO(RepairTable) = .{ .name = "grid_repair_queue_tables" },
+
+    checkpointing: ?struct {
+        /// The number of faulty_blocks with state=aborting.
+        aborting: usize,
+    } = null,
+
+    /// Guard against new block/table repairs queueing within RepairTable callbacks.
+    canceling: bool = false,
+
+    pub fn init(allocator: std.mem.Allocator, options: Options) error{OutOfMemory}!GridRepairQueue {
+        var faulty_blocks = FaultyBlocks{};
+        errdefer faulty_blocks.deinit(allocator);
+
+        try faulty_blocks.ensureTotalCapacity(
+            allocator,
+            options.blocks_max + options.tables_max * constants.lsm_table_content_blocks_max,
+        );
+
+        return GridRepairQueue{
+            .options = options,
+            .faulty_blocks = faulty_blocks,
+        };
+    }
+
+    pub fn deinit(queue: *GridRepairQueue, allocator: std.mem.Allocator) void {
+        assert(!queue.canceling);
+
+        queue.faulty_blocks.deinit(allocator);
+
+        queue.* = undefined;
+    }
+
+    /// When the queue wants more blocks than fit in a single request message, successive calls
+    /// to this function cycle through the pending BlockRequests.
+    pub fn next_batch_of_block_requests(
+        queue: *GridRepairQueue,
+        requests: []vsr.BlockRequest,
+    ) usize {
+        assert(!queue.canceling);
+        maybe(requests.len == 0);
+
+        const faults_total = queue.faulty_blocks.count();
+        if (faults_total == 0) return 0;
+        assert(faults_total > queue.faulty_blocks_repair_index);
```

Comment:
Super trivial, but let's maybe flip it around?


```zig
assert(queue.faulty_blocks_repair_index < faults_total);
```

`index < len` is how this invariant spelled most of the time (eg, in C-style for loop), comparison in the opposite direction feels odd.

At one point, I actually used to completely ban `>` and `>=` operators, to canonicalize the direction of all comparisons in a way which is consistent with the number line (this is especially nice for `lo < x && x < hi` range checks).

I now think that that goes too far, and do use `>` occasionally, but I still find canonicalizing to `<` useful, especially for index-len checks. For example, I prefer

```
if (index < len) {
} else {
}
```

and even

```
if (!(index < len)) {
}
```

to

```
if (index >= len)
```

I find it useful when invariant is stated positively and negation is pushed to the top-level.